### DISCUSSION TITLE
Quanto custa colocar uma aplicação em NodeJS no ar?

### DISCUSSION THREAD

**Discussion by wilsonneto-swe** [frontendbr/forum#1553] [Perguntas e Respostas] [answered]
Galera,
trabalho com Azure / .Net Core e também com PHP, então eu sei bem como colocar um app php é barato e como colocar uma aplicação .net no ar é caro rs

Mas e quanto ao node, quero servir uma api de ecommerce para os fronts mobile e web, meu acessos atuais são em média 1000 no dia, onde são efetuados processos na maior parte de leitura/listagem de eventos, e alguns logins e compras, a maior parte será listagem.

Mas pretendemos aumentar isso, hoje está feito em php / mysql, mas vamos migrar para implementar microserviços, cache com CQRS, MongoDB, MariaDB.

Onde e por quanto consigo servir essa api com qualidade, disponibilidade e performance?

**Comment by felipefialho** [comment]
Sugiro dar uma olhada em Serverless + AWS Lambda, a economia pode ser gigantesca, já que suas funções só são chamadas quando efetivamente tem alguém usando.

Alguns artigos:

- [The economics of serverless computing: A real-world test](https://techbeacon.com/enterprise-it/economics-serverless-computing-real-world-test)
- [DynamoDB On-Demand: When, why and how to use it in your serverless applications](https://serverless.com/blog/dynamodb-on-demand-serverless/)

**Comment by eduardojmatos** [comment]
> servir essa api com qualidade, disponibilidade e performance

Nesse caso, AWS com certeza. Se quiser algo mais simples de gerenciar, Digital Ocean é uma boa também.

Sobre usar lambda, como o @felipefialho comentou, precisa entender se a arquitetura comporta esse tipo de abordagem, apesar de ser uma boa pedida também

**Comment by ninetails** [comment]
O mais tranquilo creio que seria pegar uma infra pra rodar um container Docker. Deixar um nginx na frente, é o mais padrão (creio que seria análogo a ter um PHP em uma cloud moderna hj).

Agora se quiser aproveitar que vai refazer do zero e se preocupar em overengineering e pensando em custo, adicionando fazer o backend em FaaS (Serverless/Lambda citado pelo @felipefialho), pro front daria para pensar em alguma alternativa utilizando [JAM Stack](https://jamstack.org/) que daí pro front, o custo seria um servidor estático/CDN. Bastaria ter um serviço gerando páginas estáticas para serem _crawleadas_ e virando um SPA após o _primeiro load_ com interações com a API. Creio que os frameworks mais famosos possuem formas para exportar sites "estáticos".

**Comment by wilsonneto-swe** [comment]
Show @ninetails, vou me aprofundar sobre, indica algum serviço de servidor estático?

**Comment by delete** [comment]
> indica algum serviço de servidor estático?

@wilsonneto-dev, tem o [Netlify](https://www.netlify.com/) e o [Now](https://zeit.co/home) com planos free.

**Comment by willycamargo** [comment]
@delete acho que marcou errado rs.

@wilsonneto-dev acho que era pra ser você.

**Comment by willycamargo** [comment]
Já complementando, até onde sei o Netlify não suporta node. Apenas hospedagem de projetos estáticos.

**Comment by delete** [comment]
Marquei errado mesmo, @willycamargo , foi mal! 😄 

Sim,  Netlify só para arquivos estáticos.

**Comment by eliseumds** [comment]
@willycamargo @delete ambos Zeit Now e Netlify suportam serverless functions:

* https://zeit.co/docs/v2/serverless-functions/introduction/
* https://www.netlify.com/docs/functions/

**Comment by kavalcante** [comment]
Coloquei um projeto Node no Now.sh e o custo ficou bem show. Por problemas de negócio (teremos que fazer uma solução mais enterprise), não rolou de ficar lá.

Mas acredito que se você for seguir no front SSR com React ou Vue ou Angular, o Now.sh vai sair bem barato (e da pra começar de graça).

Quanto ao backend, utilizando Serverless ou mesmo construindo uma API mais tradicional (só recomendo o uso de Docker para facilitar deploys) o preço também não será deveras alto.

Para você ter uma ideia, hoje aqui na empresa estamos com uma média de 3 milhões de requests / mês, o que da mais ou menos 4 mil por hora ou 100.000 / dia e nossos custos de servidores estão em $50 e nosso banco (utilizamos mongodb) custa mais $60.

Não sei quais são seus custos hoje, mas para nós os nossos estão bem tranquilos e ainda iremos reduzir mais na próxima mudança de arquitetura, onde iremos centralizar tudo na AWS.

**Comment by wilsonneto-swe** [comment]
Legal galera, muito obrigado @kavalcante , @eliseumds e todos, ajudaram bastante

**Comment by diogofranco85** [comment, marked as answer]
@wilsonneto-dev  Lambda é uma boa, tem que olhar algumas particularidades, pois para validar token teria que ter uma autorizados que é mais uma lambda que se o acesso for permitido vai chamar a função a ser executada, também é bom implementar o warmup assim diminui o cold start, para front vc pode hospedar no próprio s3 da aws que tem um custo baixo também, to nessa de desenvolver uma api em node mais to pensando onde colocar também rsrsrs, custo mais baixo
